src/main/java/com/google/api/codegen/config/CollectionPatternsGenerator.java

```diff
@@ -0,0 +1,126 @@
+/* Copyright 2016 Google Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.config;
+
+import com.google.api.tools.framework.aspects.http.model.HttpAttribute;
+import com.google.api.tools.framework.aspects.http.model.HttpAttribute.FieldSegment;
+import com.google.api.tools.framework.aspects.http.model.HttpAttribute.PathSegment;
+import com.google.api.tools.framework.model.Interface;
+import com.google.api.tools.framework.model.Method;
+
+import autovalue.shaded.com.google.common.common.collect.ImmutableMap;
+
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * Class for collection config generator.
+ */
+public class CollectionPatternsGenerator implements MethodConfigGenerator {
```

Comment:
Conceptually this class is a little confusing and is not a cohesive abstraction. It has a generate(Method) method which implements the MethodConfigGenerator interface, but then it has a generateCollectionConfigs() method which needs to be called on this class directly. Maybe you need one unit that generates the name map, and separate units for generating the method-specific config and the interface-level collection config using that name map.